pkg/frontend/transport/seeker_body.go

```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package transport
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+)
+
+// ParseSeekerBodyForm will ensure that request body is a Seeker, parse the form and seek the body back to the beginning.
+func ParseSeekerBodyForm(r *http.Request) error {
+	return withSeekerBody(r, func(r *http.Request) error {
+		return r.ParseForm()
+	})
+}
+
+// withSeekerBody will ensure that request body is a Seeker, call the provided function and seek the body back to the beginning.
+func withSeekerBody(r *http.Request, f func(r *http.Request) error) (err error) {
+	if r.Body == nil {
+		return nil
+	}
+	seeker, ok := r.Body.(io.Seeker)
+	if !ok {
+		bodySeeker, err := newReadCloseSeeker(r.Body)
+		if err != nil {
+			return fmt.Errorf("can't read body: %w", err)
```

Comment:
nit: this error wrapping should be done in `newReadCloseSeeker`, that's where reading is happening.

Reply:
I wanted to avoid `newReadCloseSeeker` knowing about "body", as it's just reading an `io.Reader`.

Reply:
perhaps solution is to rename `newReadCloseSeeker` to `readIntoSeeker` and keep error wrapping here.

Reply:
Renamed: https://github.com/grafana/mimir/pull/3588/commits/6470a5992fc7c7e368a8c4516075d72ac27f32b4